lua_modules/websocket/websocket.lua

```diff
@@ -0,0 +1,496 @@
+--   -- ws.connect("ws://10.0.0.201:8989/ws")
+--   -- Example public websocket server
+--   ws.connect("ws://demos.kaazing.com/echo")
+-- end)
+-- wf.init()
+
+local m = {}
+-- m = {}
+
+-- provide your websocket url 
+m.wsUrl = "ws://10.0.0.104:8980/ws"
+
+-- private properties
+m.isInitted = false
+m.sock = nil
+m.myIp = nil
+m.isConnected = false
+m.pingTmr = nil
+
+---
+-- @name init
+-- @description Call this first with your ESP32's IP address which you
+-- can get from the wifi module.
+-- @param myip A string of your IP address like the format 10.0.0.5
+-- @returns nil
+function m.init(myip)
+  
+  if myip then m.myIp = myip end
+
+  if m.myIp == nil then
+    print("Websocket: You need to connect to wifi, or if you are, give me the IP address of this ESP32 in the init method.")
+  else 
+    print("Websocket initted. My IP: " .. m.myIp)
+    m.isInitted = true
+  end
+end
+
+m.onDataCallback = nil
+m.onConnectedCallback = nil 
+m.onDisconnectCallback = nil
+m.onPingRecvCallback = nil
+m.onPingSendCallback = nil
+m.onPongRecvCallback = nil
+---
+-- @name on
+-- @description Attach to the callback events available from this websocket library
+-- @param event_name A string of the event you want a callback on. "receive", "connection", "disconnection", "pingrecv", "pingsend", "pongrecv"
+-- @param func The callback to receive after the event
+-- @returns nil
+function m.on(method, func)
+  if method == "receive" then
+    m.onDataCallback = func 
+  elseif method == "connection" then 
+    m.onConnectedCallback = func 
+  elseif method == "disconnection" then
+    m.onDisconnectCallback = func
+  elseif method == "pingrecv" then
+    m.onPingRecvCallback = func
+  elseif method == "pingsend" then
+    m.onPingSendCallback = func
+  elseif method == "pongrecv" then
+    m.onPongRecvCallback = func
+  end 
+end
+
+---
+-- @name connect
+-- @description Ask the websocket to connect to the websocket URL
+-- @param wsurl A string of the websocket URL to connect like "ws://10.0.0.104:8980/ws"
+-- @param callback A function to get called back after a connection is established to the websocket URL
+-- @returns nil
+function m.connect(wsurl, callbackOnConnected)
+  
+  if wsurl ~= nil then m.wsUrl = wsurl end
+  if callbackOnConnected ~= nil then m.callbackOnConnected = callbackOnConnected end 
+  
+  -- Get host from websocket url 
+  local host, port, path = string.match(m.wsUrl, 'ws://(.-):(.-)/(.*)')
+  if host and port and path then
+    path = "/" .. path 
+  else
+    host, path = string.match(m.wsUrl, 'ws://(.-)/(.*)')
+    if host and path then
+      port = 80
+      path = "/" .. path 
+    else
+      host = string.match(m.wsUrl, 'ws://(.-)')
+      if host then
+        port = 80
+        path = "/"
+      else
+        print("Could not parse websocket URL?", m.wsUrl)
+      end
+      
+    end
+  end
+  
+  print("Websocket connecting to:", host, "port:", port, "path:", path)
+  
+  local body = "GET " .. path .. " HTTP/1.1\r\n"
```

Comment:
Sadly I can't use [[ or ]] because I use the chilipeppr.com/esp32 workspace to write the code and it uploads the file to the ESP32 using file.writeLine([[ line-contents ]]) technique and thus the [[ ]] in actual code messes up the upload. It would be a lot of work to solve that in ChiliPeppr. Lame reason I know, but it is, what it is.

Reply:
Well you could use [=[ and ]=] then.

Reply:
Wow, that sure gets ugly. I suppose it's more efficient, but I think this is diminishing returns. Part of why I use ESP32 is because it has tons of memory so I don't need to do whacky stuff with my code.
path = "/asdf/asdf"
host = "10.0.0.1"
local body = "GET " .. path .. " HTTP/1.1\r\n" .. [=[
Host: ]=] .. host .. "\r\n" .. [=[
Upgrade: websocket]=] .. "\r\n" .. [=[
Connection: Upgrade]=] .. "\r\n" .. [=[
Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==]=] .. "\r\n" .. [=[
Sec-WebSocket-Protocol: chat, superchat]=] .. "\r\n" .. [=[
Sec-WebSocket-Version: 13]=] .. "\r\n" .. [=[
Origin: esp32]=] .. "\r\n" .. [=[
]=] .. "\r\n"

Reply:
ah thats not what I meant.

Rather 

```
path = "/asdf/asdf"
host = "10.0.0.1"
local body = "GET " .. path .. " HTTP/1.1\r\n" .. "Host: ".. host .. [=[
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==
Sec-WebSocket-Protocol: chat, superchat
Sec-WebSocket-Version: 13
Origin: esp32

]=]
```

but not sure if the lineneds have to be \r\n in HTTP or if they are in a multiline string constant.
Maybe depends on the line end encoding of the file.

Reply:
The \r\n is required as part of the websocket standard, so you can't do the multiline the way you're describing because you'd just end up with \n per line.